### PR TITLE
Custom filter Streams

### PR DESCRIPTION
**Pull request by dahlbyk**
As [discussed](https://github.com/libgit2/libgit2sharp/pull/854#discussion_r23895973), this switches `Filter` to support reading/writing `Blob` content using `Stream`s instead of `byte[]`.

### CODE REVIEW

File: LibGit2Sharp/Core/GitBufWriteStream.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+            var length = (int)Length;
+            var bytes = new byte[length];
+            Read(bytes, 0, length);
+
+            IntPtr reverseBytesPointer = Marshal.AllocHGlobal(length);
+            Marshal.Copy(bytes, 0, reverseBytesPointer, bytes.Length);
+
+            var size = (UIntPtr)length;
+            var allocatedSize = (UIntPtr)length;
```
Comment (dahlbyk) [inline LibGit2Sharp/Core/GitBufWriteStream.cs]: In @ammeep's original implementation this was `(UIntPtr)bytes.LongLength + 1`. Based on the `AllocHGlobal()` call above this seems more correct, but I thought I'd call it out.

File: LibGit2Sharp/Core/GitBufWriter.cs [hunk 2]
```diff
@@ -1,105 +0,0 @@
...
-
-        /// <summary>
-        /// The size of the underlying stream
-        /// </summary>
-        public long Size { get; private set; }
-
-        /// <summary>
-        /// The allocated size of the underlying stream
-        /// </summary>
-        public long Allocated { get; private set; }
```
Comment (dahlbyk) [inline LibGit2Sharp/Core/GitBufWriter.cs]: Worth mentioning that this is exposed as `GitBufReadStream.Capacity`, and `Size` exposed as `GitBufReadStream.Length`.

**Comment by nulltoken** [comment]
@dahlbyk I've forced AppVeyor to build your branch. Seems we suffer from the same issue in both #946 and #948

File: LibGit2Sharp/Core/GitBufReadStream.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                   FileAccess.Read)
+        {
+            this.gitBuf = gitBuf;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            if (disposing && gitBuf != default(GitBuf))
```
Comment (ammeep) [inline LibGit2Sharp/Core/GitBufReadStream.cs]: You will need to :fire: this.

You can't dispose of the underlying git_buf. Libgit2 allocates two of these for the entire filter chain and swaps the input and output as it moves between filters in the registry. It also free's the buffers after the filter registry has been executed. This is why the tests are failing on this branch :smile:

**Comment by ammeep** [comment]
Thank you so much @dahlbyk and @nulltoken for all your hard work on this. I really appreciate the input :heart_eyes:

**Pull request merged**